### PR TITLE
fix/ldap_anonymous_bind

### PR DESCRIPTION
**Pull request by tpw56j**
Inversion of the ldap_anonymous_bind value

### CODE REVIEW

**Comment by SchoolGuy** [comment]
@tpw56j Could you somehow add tests so this doesn't happen again?

**Comment by tpw56j** [comment]
@SchoolGuy I will try to write additional tests for the parameters in this PR.

**Comment by SchoolGuy** [comment]
@tpw56j Thanks a lot!

**Review by watologo1 — APPROVED** [review]
I needed an ldap connect and used this code as reference code.
So I can confirm that the code works (I even could try an anonymous and non-anonymous connect) and that the patch is correct.

**Review by SchoolGuy — APPROVED** [review]
LGTM

**Comment by tpw56j** [comment]
@SchoolGuy To test authentication via ldap, we need an ldap server. 389-ds is available in all supported distributions. Can we include it in the docker image for testing?

**Comment by SchoolGuy** [comment]
@tpw56j Yes please do so. I had my hopes up for openLDAP but whatever works is fine. Those images are for testing, so size does not matter.

**Comment by tpw56j** [comment]
@SchoolGuy I was also hoping that it would be possible to use openLDAP, but found that RedHat stopped supporting it and removed openLDAP from its repos. I myself use openLDAP and have never used 389-ds before, so it will take me a while to figure out how to test with 389-ds in different distros.

**Comment by SchoolGuy** [comment]
@tpw56j I would prefer you get started with openLDAP then on Debian and/or openSUSE. I believe that our code should be not that different for authentication on the different distros. At least that would be my proposal.

**Comment by tpw56j** [comment]
@SchoolGuy OpenLDAP is not available only in CentOS 8. With openLDAP it will be easier for me to start, and with CentOS it will be possible to figure it out later. Therefore, I agree that I will need to start with openLDAP.

**Comment by SchoolGuy** [comment]
@tpw56j The tests are all currently failing. Is this because you forgot to add & start the LDAP Server or is this not yet finished?

**Comment by tpw56j** [comment]
@SchoolGuy I haven't added openldap to the docker image yet, so this error is expected. I will continue to work with this PR.

It was a surprise for me that for a long time I could not pass a negative test for establishing a TLS connection until I understood the difference between ldaps and starttls and that in these cases different objects should be used to set TLS parameters. All code examples I could find as well as in authentication/ldap.py used the ldap object to set TLS options. At first glance, this code looks correct, but in reality, certificates are not checked and only create the illusion of security.

Maybe add to cobbler the ability to set other TLS parameters from python-ldap? At least OPT_X_TLS_REQUIRE_CERT, OPT_X_TLS_CIPHER_SUITE, OPT_X_TLS_CRLCHECK seem important enough to me.

**Comment by SchoolGuy** [comment]
@tpw56j Okay let's do the following. You add the tests. We accept that they are failing because we know that we have a bug. Then at the same time we open a bug for this and anyone who has the time opens a followup PR for this. I see this as a separate matter from what you are trying to do with this PR.

**Comment by SchoolGuy** [comment]
If you have a better way to do it then I am open for suggestions!

**Comment by tpw56j** [comment]
@SchoolGuy Maybe I will include the necessary packages in the docker image of at least one distro and we will get a passing test in it?

Fedora requires openldap-servers, openldap-clients, python3-ldap, sscg.
Please help me add the following commands to the docker image:
```
# FQDN=`hostname`
# sscg -f -v \
     --ca-file=/etc/pki/tls/certs/ca-slapd.crt \
     --ca-key-file=/etc/pki/tls/private/ca-slapd.key \
     --ca-key-password=cobbler \
     --cert-file=/etc/pki/tls/certs/slapd.crt \
     --cert-key-file=/etc/pki/tls/private/slapd.key \
     --client-file=/etc/pki/tls/certs/ldap.crt \
     --client-key-file=/etc/pki/tls/private/ldap.key \
     --lifetime=365 \
     --hostname=$FQDN \
     --email=root@$FQDN \
# chown ldap:ldap /etc/pki/tls/{certs/slapd.crt,private/slapd.key}
# systemctl stop slapd
# rm -rf /var/lib/ldap/*  /etc/openldap/slapd.d/*
# cp /test_dir/tests/test_data/slapd.conf /etc/openldap
# systemctl start slapd
# ldapadd -Y EXTERNAL -H ldapi:/// -f /test_dir/tests/test_data/ldap_test.ldif
```

**Comment by SchoolGuy** [comment]
@tpw56j 

```Docker
ENV FQDN=`hostname`
RUN sscg -f -v \
     --ca-file=/etc/pki/tls/certs/ca-slapd.crt \
     --ca-key-file=/etc/pki/tls/private/ca-slapd.key \
     --ca-key-password=cobbler \
     --cert-file=/etc/pki/tls/certs/slapd.crt \
     --cert-key-file=/etc/pki/tls/private/slapd.key \
     --client-file=/etc/pki/tls/certs/ldap.crt \
     --client-key-file=/etc/pki/tls/private/ldap.key \
     --lifetime=365 \
     --hostname=$FQDN \
     --email=root@$FQDN \
RUN chown ldap:ldap /etc/pki/tls/{certs/slapd.crt,private/slapd.key}
RUN rm -rf /var/lib/ldap/*  /etc/openldap/slapd.d/*
RUN cp /test_dir/tests/test_data/slapd.conf /etc/openldap
RUN supervisorctl start slapd && ldapadd -Y EXTERNAL -H ldapi:/// -f /test_dir/tests/test_data/ldap_test.ldif
```

The only missing thing is not the supervisorctl config for slapd:

I found [this](https://github.com/simonswine/docker-ldapd/blob/master/supervisord.slapd.conf):

```
[program:slapd]
command=/usr/sbin/slapd -h "ldap:/// ldapi:/// ldaps:///" -g openldap -u openldap -F /srv/config -d 1 
```

You need to put that in the right directory and then you should be fine.

**Comment by tpw56j** [comment]
@SchoolGuy I was unable to add openldap to the Fedora34 container this way. And I couldn't find a way to select this container for testing.

**Comment by SchoolGuy** [comment]
We do the testing on GitHub with the openSUSE Tumbleweed container built in the openSUSE Buildservice [here](https://build.opensuse.org/package/show/home:SchoolGuy:branches:openSUSE:Templates:Images:Tumbleweed/cobbler-github-ci-devel). I will take care of the submission on OBS side once you have added all required things to the Dockerfile here in Git. For local testing we use the same dockerfile and thus you can build that container yourself like described in the Wikipage "Testing". To select the Fedora 34 container locally for testing (sadly only one-shot possible which makes it really slow) you need the `--with-tests` flag and/or `--with-system-tests` flag while building the RPMs. I saw that we don't explain this in the Wiki and I will try to add it during the day to the Wiki.

File: docker/develop/scripts/setup-openldap.sh [hunk 1]
```diff
@@ -0,0 +1,79 @@
+#!/bin/bash
+
+# Dependencies for LDAP tests
+zypper install -y openldap2 openldap2-client hostname openssl-1_1
```
Comment (SchoolGuy) [inline docker/develop/scripts/setup-openldap.sh:4]: That is a nice hack for now. Once we merge this PR, I will create a second one putting this into the image and removing this line then. Would that be okay for you?

Comment (tpw56j) [reply docker/develop/scripts/setup-openldap.sh:4]: Ok

**Comment by SchoolGuy** [comment]
@tpw56j You are aware that https://pypi.org/project/ldap/ is a placeholder which we do not want probably? Or is this an intended change?

**Comment by tpw56j** [comment]
@SchoolGuy cobbler/modules/authentication/ldap.py does not use ldap3, so I replaced ldap3 with ldap everywhere. I will remove https://pypi.org/project/ldap.

**Comment by SchoolGuy** [comment]
@tpw56j Oh that sounds like a lot of tech-debt. Do you need help with something or should I just monitor your progress and review?

**Comment by tpw56j** [comment]
@SchoolGuy Thanks for your willingness to help, I'll try to finish the ldap-related code changes myself.

**Comment by tpw56j** [comment]
@SchoolGuy My branch is missing Debian 11, so I cannot make changes there. And in system-test there is an error not related to my changes (dhcpd).

**Comment by SchoolGuy** [comment]
@tpw56j You could rebase the branch on `master`, right? But yeah the error on the system-tests is fine. I am aware of that.

**Comment by SchoolGuy** [comment]
So building works fine but installing does not. Does Debian 11 not have the required packages?

**Review by SchoolGuy — APPROVED** [review]
LGTM even with the new changes!

**Comment by SchoolGuy** [comment]
@tpw56j Is this fully ready in your eyes?

**Comment by tpw56j** [comment]
@SchoolGuy So far I don't like this added code:
```
        if tls_reqcert:
            if tls_reqcert == "never":
                ldaps_tls.set_option(ldap.OPT_X_TLS_REQUIRE_CERT, ldap.OPT_X_TLS_NEVER)
            elif tls_reqcert == "allow":
                ldaps_tls.set_option(ldap.OPT_X_TLS_REQUIRE_CERT, ldap.OPT_X_TLS_ALLOW)
            elif tls_reqcert == "demand":
                ldaps_tls.set_option(ldap.OPT_X_TLS_REQUIRE_CERT, ldap.OPT_X_TLS_DEMAND)
            elif tls_reqcert == "hard":
                ldaps_tls.set_option(ldap.OPT_X_TLS_REQUIRE_CERT, ldap.OPT_X_TLS_HARD)
            else:
                api_handle.logger.error("authn_ldap: ldap_tls_reqcert choices include: %s" %
                                        "never, allow, demand, hard")
```
I think I figured out how to change it in this PR and I'll try to do it.

And we also do not have a mechanism for validating a fixed list of supported values in the settings, such as, for example, ldap_tls_reqcert or default_template_type, default_virt_type. But this change would probably be better done in a separate PR.

**Comment by SchoolGuy** [comment]
@tpw56j Thanks a lot for the status update! Take your time and don't feel rushed!

**Comment by tpw56j** [comment]
@SchoolGuy PR is ready to review

**Review by SchoolGuy — APPROVED** [review]
System-Tests are broken currently but this is not the concern of this PR. Since we have new tests which cover most/some of the new code and this was previously untested, I will accept the new PR. Thanks for your work @tpw56j !

**Pull request merged**